icu4c/source/common/brkeng.cpp

```diff
@@ -299,6 +306,91 @@ ICULanguageBreakFactory::loadDictionaryMatcherFor(UScriptCode script) {
     return nullptr;
 }
 
+
+URegistryKey ICULanguageBreakFactory::addExternalEngine(
+    ExternalBreakEngine* engine, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+    ensureEngines(status);
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+    ExternalLanguageBreakEngine* wrapper = new ExternalLanguageBreakEngine(engine);
+    fEngines->push(wrapper, status);
+    return (URegistryKey)wrapper;
+}
+
+UBool ICULanguageBreakFactory::removeExternalEngine(
+    URegistryKey key, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return false;
+    }
+    ensureEngines(status);
+    if (U_FAILURE(status)) {
+        return false;
+    }
+    int32_t found = fEngines->search((ExternalLanguageBreakEngine*)key);
+    if (found < 0) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return false;
+    }
+    fEngines->removeElementAt(found);
+    return true;
+}
+
+ExternalLanguageBreakEngine::ExternalLanguageBreakEngine(
+    ExternalBreakEngine* engine) : delegate(engine) {
+}
+
+ExternalLanguageBreakEngine::~ExternalLanguageBreakEngine() {
+    delete delegate;
+}
+
+UBool ExternalLanguageBreakEngine::handles(UChar32 c, const char* locale) const {
+    return delegate->isFor(c, locale);
+}
+
+int32_t ExternalLanguageBreakEngine::findBreaks(
+    UText *text,
+    int32_t startPos,
+    int32_t endPos,
+    UVector32 &foundBreaks,
+    UBool /* isPhraseBreaking */,
+    UErrorCode &status) const {
+    if (U_FAILURE(status)) return 0;
+    int32_t result = 0;
+
+    // Find the span of characters included in the set.
+    //   The span to break begins at the current position in the text, and
+    //   extends towards the start or end of the text, depending on 'reverse'.
+
+    utext_setNativeIndex(text, startPos);
+    int32_t start = (int32_t)utext_getNativeIndex(text);
+    int32_t current;
+    int32_t rangeStart;
+    int32_t rangeEnd;
+    UChar32 c = utext_current32(text);
+    while((current = (int32_t)utext_getNativeIndex(text)) < endPos && delegate->handles(c)) {
+        utext_next32(text);         // TODO:  recast loop for postincrement
+        c = utext_current32(text);
+    }
+    rangeStart = start;
+    rangeEnd = current;
+    int32_t beforeSize = foundBreaks.size();
+    int32_t additionalCapacity = rangeEnd - rangeStart + 1;
+    // enlarge to contains (rangeEnd-rangeStart+1) more items
+    foundBreaks.ensureCapacity(beforeSize+additionalCapacity, status);
+    if (U_FAILURE(status)) return 0;
+    foundBreaks.setSize(beforeSize + beforeSize+additionalCapacity);
+    result = delegate->fillBreak(text, rangeStart, rangeEnd, foundBreaks.getBuffer(),
```

Comment:
`foundBreaks.getBuffer()`, should be `foundBreaks.getBuffer()+beforeSize` ?
Looks like the code is setting up  to append, then writes at the start of `foundBreaks`.